bonus/hkd.md

```diff
@@ -0,0 +1,367 @@
+# HKD - Higher-Kinded Data
+
+This file is literate haskell. If you have
+[`markdown-unlit`](https://hackage.haskell.org/package/markdown-unlit)
+installed, you can extract the code by running the following command:
+
+```
+$ markdown-unlit -h mtl.md mtl.md mtl.hs
+```
+
+This will allow you to work through the exercises using GHCi.
+
+In this document, we will explore a pattern called "higher-kinded
+data", which is when a data structure is parameterised by a
+`Functor`. This is becoming increasingly common in libraries like
+[`waargonaut`](https://hackage.haskell.org/package/waargonaut) (which
+we used for JSON encoding/decoding),
+[`dependent-sum`](https://hackage.haskell.org/package/dependent-sum),
+and
+[`dependent-map`](https://hackage.haskell.org/package/dependent-map),
+so it is worth knowing about.
+
+
+<details>
+<summary>Extensions and Imports</summary>
+
+```haskell
+{-# OPTIONS_GHC -Wall -Wno-unused-imports -Wno-unused-matches #-}
+
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE NoStarIsType #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+import Data.Functor.Compose (Compose(..))
+import Data.Functor.Identity (Identity(..))
+import qualified Data.Semigroup as Semigroup
+import Data.Kind (Type)
+import Data.Word (Word16)
+```
+
+</details>
+
+
+## Generalising `Conf`
+
+Consider the configuration types from level 7:
+
+```haskell
+newtype Port = Port Word16
+
+data Conf = Conf
+  { port       :: Port
+  , dbFilePath :: FilePath
+  }
+
+data PartialConf = PartialConf
+  { pcPort       :: Maybe (Semigroup.Last Port)
+  , pcDBFilePath :: Maybe (Semigroup.Last FilePath)
+  }
+```
+
+There's repeated structure here. We can generalise over the both of
+these structures by introducing a `Functor`-kinded type parameter:
```

Comment:
"But one is `Maybe`, the other is not, that doesn't seem repeated?"
"It's not necessarily obvious, but we know that `Maybe` is a `Functor, yes? Without changing the meaning of `Config` we're able to add `Identity` to `Port` and `FilePath`. Giving us `Identity Port` etc... and `Identity` has a `Functor` instance. So without changing the meaning of either of our types, we've found a way to abstract away the whole structure. Leaving us with the truly important information.. Which is we can't be sure we have all of the config, ala `Maybe`, or we've a complete config, ala `Identity`. 

... or something to that effect... Remembering that you can _introduce_ structure like `Identity` is super non-obvious to me, and I expect I'm not alone in that. Be good to try to flesh it out a bit.